### PR TITLE
Merchant Item Stats and Nav Partials

### LINKED ISSUE #28: Merchant Items Index: Top Item's Best Day
When I visit the items index page
Then next to each of the 5 most popular items I see the date with the most sales for each item.
And I see a label “Top selling date for <item name> was <date with most sales>"

Note: use the invoice date. If there are multiple days with equal number of sales, return the most recent day.

### LINKED ISSUE #29: Merchant Items Index: 5 most popular items
As a merchant
When I visit my items index page
Then I see the names of the top 5 most popular items ranked by total revenue generated
And I see that each item name links to my merchant item show page for that item
And I see the total revenue generated next to each item name

Notes on Revenue Calculation:
- Only invoices with at least one successful transaction should count towards revenue
- Revenue for an invoice should be calculated as the sum of the revenue of all invoice items
- Revenue for an invoice item should be calculated as the invoice item unit price multiplied by the quantity (do not use the item unit price)

### LINKED ISSUE #52: Create partial view for site header and nav for use across all pages

### PR DESCRIPTION
**Pull request by mkrumholz**
## Changes

- Created methods to display merchant item revenue and best revenue days
- Built 2 partials for header navs for admin and merchant pages

## Checklist
- [x] Tests written where required
- [x] All tests passing
- [x] SimpleCov at 100%

closes #28 
closes #29 
closes #52

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -13,6 +13,17 @@ def self.unshipped_items
     .order('invoices.created_at asc')
   end
 
+  def self.highest_revenue_date(item_id)
```
Comment (Sidarth20) [inline app/models/invoice.rb:16]: Great job with this query!

File: app/models/invoice.rb [hunk 2]
```diff
@@ -13,6 +13,17 @@ def self.unshipped_items
...
+    joins(:invoice_items)
+    .joins('right join transactions on transactions.invoice_id=invoices.id')
+    .select(invoices: :updated_at)
+    .where(invoice_items: {item_id: item_id}, transactions: {result: 1}, invoices: {status: 1})
+    .order(Arel.sql('sum(invoice_items.quantity * invoice_items.unit_price) desc, invoices.updated_at desc'))
```
Comment (Sidarth20) [inline app/models/invoice.rb:21]: @mkrumholz potential refactor to calculate the revenue in the `.select `with an `AS revenue` & then call it in the `.order` without needing the `AREL.sql`

File: app/models/item.rb [hunk 3]
```diff
@@ -19,6 +19,22 @@ def self.disabled_only
     where(enabled: false)
   end
 
+  def self.top_5_by_revenue 
+    joins(:invoice_items)
```
Comment (Sidarth20) [inline app/models/item.rb:23]: @mkrumholz Could you do `.joins(:invoice_items {invoices: :transactions})`?

File: app/models/item.rb [hunk 4]
```diff
@@ -19,6 +19,22 @@ def self.disabled_only
...
+    .joins('right join invoices on invoices.id=invoice_items.invoice_id')
+    .joins('right join transactions on transactions.invoice_id=invoices.id')
+    .select('items.*, sum(invoice_items.quantity * invoice_items.unit_price) as revenue')
+    .where(transactions: {result: 1})
```
Comment (Sidarth20) [inline app/models/item.rb]: @mkrumholz Could you do `.where(transactions.result => 1, invoices.status =>1)`?

File: app/views/partials/_admin_nav.html.erb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<nav>
+  <h2>Admin Dashboard</h2>
+  <ul>
+    <li><%= link_to 'Dashboard', admin_index_path %></li>
```
Comment (Sidarth20) [inline app/views/partials/_admin_nav.html.erb:4]: nice job with the partials!

File: app/models/invoice.rb (hunk 2 above)
Comment (mkrumholz) [reply app/models/invoice.rb:21]: @Sidarth20 I just went and tried this, but it causes an error. Honestly not sure why, but it won't let me use the alias in my order clause.

File: spec/features/merchants/items/index_spec.rb [hunk 6]
```diff
@@ -93,4 +124,71 @@
 
     expect(current_path).to eq "/merchants/#{@merchant.id}/items/new"
   end
+
+  it 'lists the top 5 best-selling items by revenue generated' do
+    visit "/merchants/#{@merchant.id}/items"
+
+    within "section#popular" do  
+      expect(@item_1.name).to appear_before @item_2.name
```
Comment (Sidarth20) [inline spec/features/merchants/items/index_spec.rb:132]: Great use of orderly & within testing!

File: spec/features/merchants/items/index_spec.rb [hunk 7]
```diff
@@ -3,10 +3,35 @@
...
+
     @item_1 = @merchant.items.create!(name: 'Audrey II', description: 'Large, man-eating plant', unit_price: '100000000', enabled: true)
     @item_2 = @merchant.items.create!(name: 'Bouquet of roses', description: '12 red roses', unit_price: '1900', enabled: true)
     @item_3 = @merchant.items.create!(name: 'Orchid', description: 'Purple, 3 inches', unit_price: '2700', enabled: false)
     @item_4 = @merchant.items.create!(name: 'Echevaria', description: 'Peacock varietal', unit_price: '3100', enabled: true)
+    @item_5 = @merchant.items.create!(name: 'Fourth item', description: '4th best', unit_price: '26400', enabled: true)
+    @item_6 = @merchant.items.create!(name: 'Fifth item', description: '5th best', unit_price: '2400', enabled: true)
+    @item_7 = @merchant.items.create!(name: 'Sixth item', description: '6th best', unit_price: '50', enabled: true)
+
+    @invoice_1 = @customer.invoices.create!(status: 1, updated_at: Date.parse("2021-03-01")) # is successful and paid
```
Comment (Sidarth20) [inline spec/features/merchants/items/index_spec.rb:16]: Thanks for refactoring this!!!

File: spec/models/invoice_spec.rb [hunk 8]
```diff
@@ -47,7 +44,46 @@
         expect(Invoice.unshipped_items.ids.include?(@invoice_5.id)).to eq(false)
       end
     end
+
+    describe '#highest_revenue_date' do
+      before :each do
+        @merchant = Merchant.create!(name: "Little Shop of Horrors")
+        @customer = Customer.create!(first_name: 'Audrey', last_name: 'I')
+
```
Comment (Sidarth20) [inline spec/models/invoice_spec.rb:52]: Nice work on the refactor!!!

File: spec/models/item_spec.rb [hunk 9]
```diff
@@ -12,63 +12,89 @@
...
+        expect(Item.top_5_by_revenue).to include @item_4
+        expect(Item.top_5_by_revenue).to include @item_5
+        expect(Item.top_5_by_revenue).to include @item_6
+        expect(Item.top_5_by_revenue).to_not include @item_3
+        expect(Item.top_5_by_revenue).to_not include @item_7
+      end
+    end 
+
     describe '.ready_to_ship' do
       it 'returns only the items where invoice_item status = packaged & sorted by oldest to newest' do
```
Comment (Sidarth20) [inline spec/models/item_spec.rb:79]: Thank you for your hard work on the refactor!!!

File: app/models/item.rb (hunk 3 above)
Comment (mkrumholz) [reply app/models/item.rb:23]: Ah yea, it looks like that works with some minor adjustments -> `.joins(invoice_items: {invoice: :transactions})`

Comment (mkrumholz) [reply app/models/item.rb:23]: Nice!

**Review by Sidarth20 — APPROVED** [review]

File: app/models/item.rb (hunk 4 above)
Comment (mkrumholz) [reply app/models/item.rb]: Will update -- you don't need the hash rockets, though. Just plain `transactions: {result: 1}, invoices: {status: 1}`

Comment (mkrumholz) [reply app/models/item.rb]: I just don't like hash rockets. Rockets belong in space not code 🚀

**Pull request merged**